### PR TITLE
Add the runtime-manager design doc

### LINKED ISSUE #144: [proposal] runtime-manager manunal
**What is your proposal**:

**Why is this needed**:

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by honpey**
Signed-off-by: honpey <honpey@gmail.com>

### Ⅰ. Describe what this PR does
fixes #144

### CODE REVIEW

File: docs/design-archive/runtime-manager-design-doc.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        * [Examples for Runtime Plugins](#examples-for-runtime-plugins)
+    * [Installation](#installation)
+
+## Summary
+
+RuntimeManager acts as a proxy between kubelet and containerd(dockerd under dockershim scenario), which is designed to
+intercept CRI request, and apply some resource management policies, such as setting different cgroup parameters by pod
+priorities under hybrid scenario, applying new isolation policies for new linux kernel and latest cpu architecture etc.
+
+There are two components involved, RuntimeManager and RuntimePlugins.<br><br>
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:26]: <br> could be replaced with empty line

File: docs/design-archive/runtime-manager-design-doc.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+If no RuntimePlugins registered, RuntimeManager would become a transparent proxy between kubelet and containerd.
+
+## Architecture
+
+![image](../images/runtime-manager-design.svg)
+
+There are mainly 4 components for RuntimeManager.
+
+### Cri Server
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:60]: CRI Server

File: docs/design-archive/runtime-manager-design-doc.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+RuntimeManager, RuntimeManager would decide to transfer request to backend containerd or discard request according to
+plugins' response.
+
+If no RuntimePlugins registered, RuntimeManager would become a transparent proxy between kubelet and containerd.
+
+## Architecture
+
+![image](../images/runtime-manager-design.svg)
+
+There are mainly 4 components for RuntimeManager.
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:58]: There are 4 main components for RuntimeManager:

File: docs/design-archive/runtime-manager-design-doc.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+### Cri Server
+
+as a proxy between kubelet and containerd, RuntimeManager acts as a CRI server for kubelet(http server under dockershim
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:62]: `As` a proxy

File: docs/design-archive/runtime-manager-design-doc.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+scenario). It should intercept all request from kubelet, and generate protocols for talking with plugins before and
+after talking with backend containerd(dockerd)
+
+### Plugins Manager
+
+PluginsManager is in charge of parsing registered plugin info from /etc/runtime/hookserver.d dynamically. examples for
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:68]: `Examples` for
and btw, the example is missing.

File: docs/design-archive/runtime-manager-design-doc.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+plugin register file:
+
+### Runtime Dispatcher
+
+RuntimeDispatcher is designed to manage communications with plugins.
+
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. For StartContainer hook
+example, when StartContainer request comes, there exists only containerID in CRI StartContainerRequest, which is not
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:78]: `when StartContainer request comes, there exists only containerID in CRI StartContainerRequest` -> `there is only containerID in CRI StartContainerRequest`

File: docs/design-archive/runtime-manager-design-doc.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+enough for plugins to adapt policies since plugins may not store pod/container info(such as meta, priority) locally. So
+RuntimeManager should store pod/container info during RunPodSandbox/CreateContainer Stage. when StartContainer request
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:80]: `When` StartContainer request

File: docs/design-archive/runtime-manager-design-doc.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+```
+{
+    "remote-endpoint": "/var/run/koordlet-runtimehookserver.sock",
+    "failure-policy": "Ignore",
+    "runtime-hooks": [
+        "PreRunPodSandbox"
+    ]
+}
+```
+there are mainly 3 fields involved:
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:99]: `There` are

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

File: docs/design-archive/runtime-manager-design-doc.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        * [How to Register Plugins](#how-to-register-plugins)
+        * [Protocols between RuntimeManager and Plugins](#protocols-between-runtimemanager-and-plugins)
+        * [Examples for Runtime Plugins](#examples-for-runtime-plugins)
+    * [Installation](#installation)
+
+## Summary
+
+RuntimeManager acts as a proxy between kubelet and containerd(dockerd under dockershim scenario), which is designed to
+intercept CRI request, and apply some resource management policies, such as setting different cgroup parameters by pod
+priorities under hybrid scenario, applying new isolation policies for new linux kernel and latest cpu architecture etc.
```
[the remark this answers is not in the document]
Comment (honpey) [reply docs/design-archive/runtime-manager-design-doc.md]: fixed to 'latest Linux kernel, CPU architecture, and etc.'

File: docs/design-archive/runtime-manager-design-doc.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+PluginsManager is in charge of parsing registered plugin info from /etc/runtime/hookserver.d dynamically. examples for
+plugin register file:
+
+### Runtime Dispatcher
+
+RuntimeDispatcher is designed to manage communications with plugins.
+
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. For StartContainer hook
```
[the remark this answers is not in the document]
Comment (honpey) [reply docs/design-archive/runtime-manager-design-doc.md]: > `For StartContainer hook example` -> `Take ... for example`

done

File: docs/design-archive/runtime-manager-design-doc.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+since plugins may not store pod/container info(such as meta, priority) locally. So RuntimeManager should store pod/container
+info during RunPodSandbox/CreateContainer Stage. When StartContainer request comes, RuntimeManager can get pod/container info
+by containerID, and then call plugins with pod/container info.<br>
+With store, there would be pod/container info everytime RuntimeManager calls plugins, so there is no need for plugins to
+store pod/container info exceptionally, plugins can be designed as stateless.
+
+## Runtime Plugins
+
+### How to Register Plugins
+All the plugin config files should be put to /etc/runtime/hookserver.d with .json suffix, for example:
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:88]: enclose `/etc/runtime/hookserver.d` and `.json` in backticks (`)

File: docs/design-archive/runtime-manager-design-doc.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+/etc/runtime/hookserver.d/koordlet.json
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:89]: include this line in the code block below. for example:
```
$ cat /etc/runtime/hookserver.d/koordlet.json
{
    "remote-endpoint": "/var/run/koordlet-runtimehookserver.sock",
    "failure-policy": "Ignore",
    "runtime-hooks": [
        "PreRunPodSandbox"
    ]
}
```

File: docs/design-archive/runtime-manager-design-doc.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+RuntimeDispatcher is designed to manage communications with plugins.
+
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. Take StartContainer hook
+for example, there exists only containerID in CRI StartContainerRequest, which is not enough for plugins to adapt policies
+since plugins may not store pod/container info(such as meta, priority) locally. So RuntimeManager should store pod/container
+info during RunPodSandbox/CreateContainer Stage. When StartContainer request comes, RuntimeManager can get pod/container info
+by containerID, and then call plugins with pod/container info.<br>
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:81]: replace <br> with empty line.

File: docs/design-archive/runtime-manager-design-doc.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+- Provide interface for new isolation features which are not supported by CRI.
+
+## Components
+
+### RuntimeManager
+
+RuntimeManager is in charge of intercepting request during pod's lifecycle, such as RunPodSandbox, CreateContainer etc.,
+and then calling RuntimePlugins to do resource isolation policies before transferring request to backend containerd(
+dockerd)
+and after transferring response to kubelet. RuntimeManager provides an isolation-policy-execution framework which allows
```
Comment (jasonliu747) [inline docs/design-archive/runtime-manager-design-doc.md:42]: why starts a new line here?

**Review by jasonliu747 — COMMENTED** [review]
/lgtm
others PTAL 
/cc @hormes @saintube @zwzhang0107 @stormgbs @ZYecho

File: docs/design-archive/runtime-manager-design-doc.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        * [How to Register Plugins](#how-to-register-plugins)
+        * [Protocols between RuntimeManager and Plugins](#protocols-between-runtimemanager-and-plugins)
+        * [Examples for Runtime Plugins](#examples-for-runtime-plugins)
+    * [Installation](#installation)
+
+## Summary
+
+RuntimeManager acts as a proxy between kubelet and containerd(dockerd under dockershim scenario), which is designed to
+intercept CRI request, and apply some resource management policies, such as setting different cgroup parameters by pod
+priorities under hybrid scenario, applying new isolation policies for latest Linux kernel, CPU architecture, and etc.
```
Comment (hormes) [inline docs/design-archive/runtime-manager-design-doc.md:24]: under hybrid -> under hybrid workloads orchestration scenario

File: docs/design-archive/runtime-manager-design-doc.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+There are two components involved, RuntimeManager and RuntimePlugins.
+
+![image](../images/runtime-manager-architecture.svg)
+
+## Goals
+
+- Enhance resource management for scenario like hybridization.
```
Comment (hormes) [inline docs/design-archive/runtime-manager-design-doc.md:32]: Enhance resource management for QoS based Scheduling

File: docs/design-archive/runtime-manager-design-doc.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+### Store
+
+As a proxy, RuntimeManager had better be designed as stateless, but sometimes it does NOT work. Take StartContainer hook
+for example, there exists only containerID in CRI StartContainerRequest, which is not enough for plugins to adapt policies
+since plugins may not store pod/container info(such as meta, priority) locally. So RuntimeManager should store pod/container
+info during RunPodSandbox/CreateContainer Stage. When StartContainer request comes, RuntimeManager can get pod/container info
+by containerID, and then call plugins with pod/container info.
+
+With store, there would be pod/container info everytime RuntimeManager calls plugins, so there is no need for plugins to
+store pod/container info exceptionally, plugins can be designed as stateless.
```
Comment (hormes) [inline docs/design-archive/runtime-manager-design-doc.md:83]: Add a consideration about performance, store is in memory and does not generate external io to disk.

Comment (honpey) [reply docs/design-archive/runtime-manager-design-doc.md:84]: > Add a consideration about performance, store is in memory and does not generate external io to disk.

changed to: 
Considering performance, store locates in memory and does not generate external io to disk.

File: docs/design-archive/runtime-manager-design-doc.md (hunk 15 above)
Comment (honpey) [reply docs/design-archive/runtime-manager-design-doc.md]: > under hybrid -> under hybrid workloads orchestration scenario

fixed

File: docs/design-archive/runtime-manager-design-doc.md (hunk 16 above)
Comment (honpey) [reply docs/design-archive/runtime-manager-design-doc.md]: > Enhance resource management for QoS based Scheduling

fixed

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Pull request merged**